src/base/bittorrent/tracker.h

```diff
@@ -48,54 +53,58 @@ namespace BitTorrent
 {
     struct Peer
     {
-        QHostAddress ip;
         QByteArray peerId;
-        int port;
+        QHostAddress address;  // reported by socket
+        QByteArray claimedAddress;  // self claimed by peer
+        ushort claimedPort = 0;  // self claimed by peer
+        bool isSeeder = false;
 
-        bool operator!=(const Peer &other) const;
-        bool operator==(const Peer &other) const;
-        QString uid() const;
-        lt::entry toEntry(bool noPeerId) const;
-    };
+        lt::entry::string_type addressByteString;
```

Comment:
>I would use this field instead of uid().

That wouldn't be correct.

>Also "endpoint" is usual name for ip+port pair.

ok.

Reply:
>That wouldn't be correct.

Why?

Reply:
>Why?

It is possible that `uid()` and `endpoint` will have different values isn't? For example that `claimedAddress` might be a domain name/path and thus I consider the current `uid()` way is correct/more suitable.

Reply:
I decided to add a `QByteArray uniqueID;` variable so qHash have something to use directly.

Reply:
>For example that claimedAddress might be a domain name/path 

Yes.
>and thus I consider the current uid() way is correct/more suitable.

Why? Isn't address provided by peer itself considered to be more actual/unique than the one we get from socket?

Reply:
>I decided to add a QByteArray uniqueID; variable so qHash have something to use directly.

>Update: Still decided to use a function instead.

As for this and the other optimizations I suggested earlier, you just have to figure out how frequently it's used in real life. If frequently, it is preferable to have it all serialized, if not, you can do serialization for each processed request.

Reply:
>Why? Isn't address provided by peer itself considered to be more actual/unique than the one we get from socket?

I believe this is already addressed in the last update. Now `uniqueID()` is using `ip` field + port number.
`ip` and `endpoint` fields are still different as `ip` can contain domain name while `endpoint` will only contain IP address + port (if claimed address is an IP address otherwise socket address). You might wonder why I don't resolve domain name and use its IP address, well that is because resolving DNS is very time consuming, I rather pass it as-is to other peers and let them resolve it themselves.

>As for this and the other optimizations I suggested earlier, you just have to figure out how frequently it's used in real life. 

(Without measurements) I believe it is better to precompute them, it will do much less work when there are many peers.

Reply:
> I believe it is better to precompute them, it will do much less work when there are many peers.

But uid is not still precomputed. Isn't it accessed too frequently?

Reply:
>>I believe it is better to precompute them, it will do much less work when there are many peers.
>
>But uid is not still precomputed. Isn't it accessed too frequently?

I was talking about the `ip` field specifically. Yes, uid is not precomputed, however most of it is composed of `ip` field and uid() merely adds a colon and a number which all of them are in ascii range so the conversion and appending can be very fast, at least fast enough for hash use IMO.
As for frequency, I don't believe it is frequent enough to allocate it a specific field in Peer class.